File: app/lib/services/cloud_transfer.dart
import 'cancel_token.dart';

typedef OnTransferProgress = void Function(int transferred, int total);

class CloudUploadResult {
  final String key;
  final String fileName;
  CloudUploadResult({required this.key, required this.fileName});
}

class CloudDownloadResult {
  final String filePath;
  final int totalBytes;
  CloudDownloadResult({required this.filePath, required this.totalBytes});
}

abstract class CloudTransferService {
  Future<CloudUploadResult> upload({
    required String fileName,
    required int fileSize,
    String? filePath,
    List<int>? bytes,
    String? contentType,
    CancelToken? cancelToken,
    OnTransferProgress? onProgress,
  });

  Future<CloudDownloadResult> download({
    required String key,
    required String savePath,
    CancelToken? cancelToken,
    OnTransferProgress? onProgress,
    int? lastModifiedMs,
  });
}

File: web/src/lib/services/cloudTransfer.ts
export type OnTransferProgress = (transferred: number, total: number) => void;

export type CloudUploadResult = {
  key: string;
  fileName: string;
};

export type CloudDownloadResult = {
  blob: Blob;
  totalBytes: number;
};

export interface CloudTransferService {
  upload(
    file: File,
    onProgress?: OnTransferProgress,
    abortSignal?: AbortSignal,
  ): Promise<CloudUploadResult>;

  download(
    key: string,
    onProgress?: OnTransferProgress,
  ): Promise<CloudDownloadResult>;
}
